### PR TITLE
Make .NET Core runner load dependencies correctly when running benchmarks

### PR DESCRIPTION
**Pull request by heynickc**
Per #192, this WIP looks for dependencies for the benchmark assembly and loads them along with the benchmark assembly.

### CODE REVIEW

**Comment by Aaronontheweb** [comment]
@heynickc mono build failure?

**Comment by heynickc** [comment]
@Aaronontheweb should pass End2End tests now, but might need review/suggestions on how to do this better.

**Comment by Aaronontheweb** [comment]
@heynickc instead of referencing the Akka.Actor source directly for this test (I know that's where the reproduction of where the issue came from), could you just reference the NuGet packages instead? This one is a doosey to review at the moment with ~400 files changed :p

**Comment by heynickc** [comment]
@Aaronontheweb Ok, I was going to use NuGet but I needed the newest .NET Core version of Akka to test against.  Could also do a submodule maybe?  Not sure if we're publishing v1.3 Akka branch to MyGet yet but maybe we should sometime.

**Comment by Aaronontheweb** [comment]
> Ok, I was going to use NuGet but I needed the newest .NET Core version of Akka to test against

Argh.... right.... Well, in the short run, let's just commit the .NET Core-compiled binaries in with the source and add an issue to fetch them properly from NuGet later. Sound good? Probably easier than including the .NET Core nightly build for Akka right now anyway.

**Comment by Aaronontheweb** [comment]
@heynickc test timeout

**Comment by Aaronontheweb** [comment]
@heynickc mono build failure

**Comment by heynickc** [comment]
this PR will also resolve #182 and #3

File: build.sh [hunk 1]
```diff
@@ -85,6 +85,11 @@ if [ ! -f "$FAKE_EXE" ]; then
 fi
 
 ###########################################################################
+# WORKAROUND FOR MONO
+###########################################################################
+export FrameworkPathOverride=/usr/lib/mono/4.5/
```
Comment (Aaronontheweb) [inline build.sh]: @heynickc what issue did this solve on Mono?

File: src/NBench/Sdk/Compiler/AssemblyRuntimeLoader.cs [hunk 2]
```diff
@@ -43,15 +48,53 @@ public static bool CanFindAssembly(string assemblyPath)
...
         /// <param name="assemblyPath">The path to an assembly</param>
         /// <returns>The assembly at the specified location</returns>
-        public static Assembly LoadAssembly(string assemblyPath)
+        public static Assembly[] LoadAssembly(string assemblyPath)
         {
 #if CORECLR
-            return AssemblyLoadContext.Default.LoadFromAssemblyPath(assemblyPath);
+            //var assemblies = ReflectionDiscovery.GetAssemblies(); // TODO: net45 AssemblyResolve delegate has potential here
+            //return assemblies;
+            AssemblyLoadContext.Default.Resolving += (assemblyLoadContext, assemblyName) => DefaultOnResolving(assemblyLoadContext, assemblyName, assemblyPath);
```
Comment (Aaronontheweb) [inline src/NBench/Sdk/Compiler/AssemblyRuntimeLoader.cs]: lol... way to go .NET Core... Rename an existing API and slightly change it semantics without altering its purpose.... Shuffling deck chairs... Anyway, good on you for figuring these plat differences out.

File: src/NBench/Sdk/Compiler/IDiscovery.cs [hunk 3]
```diff
@@ -30,7 +30,7 @@ public interface IDiscovery
         /// </summary>
         /// <param name="targetAssembly">The assembly we're going to scan for benchmarks.</param>
         /// <returns>A list of <see cref="Benchmark"/>s we can run based on the classes found inside <paramref name="targetAssembly"/>.</returns>
-        IEnumerable<Benchmark> FindBenchmarks(Assembly targetAssembly);
+        IEnumerable<Benchmark> FindBenchmarks(Assembly[] targetAssembly);
```
Comment (Aaronontheweb) [inline src/NBench/Sdk/Compiler/IDiscovery.cs]: What's the purpose of this change? Allowing the NBench runner to test against multiple assemblies in the same run of `NBench.Runner.exe`?

File: src/NBench/Sdk/Compiler/ReflectionDiscovery.Configurators.cs [hunk 4]
```diff
@@ -89,7 +89,7 @@ public static IEnumerable<Type> LoadAllTypeConfigurators()
                 types.Where(IsConfigurationType);
         }
 #if CORECLR
-        private static Assembly[] GetAssemblies()
+        public static Assembly[] GetAssemblies()
```
Comment (Aaronontheweb) [inline src/NBench/Sdk/Compiler/ReflectionDiscovery.Configurators.cs]: Why `public` for .NET Core?

File: tests/NBench.Tests.End2End/NBenchIntegrationTest.WithDependencies.cs [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+        public NBenchIntregrationTestWithDependencies(ITestOutputHelper output)
+        {
+            _output = output;
+        }
+
+        [Fact]
+        public void ShouldPassAllBenchmarks()
+        {
+            if (!TestRunner.IsMono) // this spec currently hits a runtime exception with Mono: https://bugzilla.xamarin.com/show_bug.cgi?id=43291
```
Comment (Aaronontheweb) [inline tests/NBench.Tests.End2End/NBenchIntegrationTest.WithDependencies.cs]: Good times. Can you file an issue on the NBench repo to track this one on the Xamarin bug-tracker?

File: tests/NBench.Tests.Performance.WithDependencies/Actor/ActorThroughputSpec.cs [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+            {
+                _actorBasePatternMatchActorRef.Tell(dataExample);
+                _actorBasePatternMatchActorRef.Tell(intExample);
+                _actorBasePatternMatchActorRef.Tell(stringExample);
+                ++i;
+            }
+            _resetEvent.Wait(); //wait up to a second
+        }
+
+        //[PerfBenchmark(
```
Comment (Aaronontheweb) [inline tests/NBench.Tests.Performance.WithDependencies/Actor/ActorThroughputSpec.cs]: Why is all of this stuff commented out? Just delete it if we aren't going to use it IMHO

File: tests/NBench.Tests.Performance/NBench.Tests.Performance.csproj [hunk 7]
```diff
@@ -1,7 +1,7 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFrameworks>net452;netcoreapp1.1</TargetFrameworks>
+    <TargetFrameworks>net452;netstandard1.6</TargetFrameworks>
```
Comment (Aaronontheweb) [inline tests/NBench.Tests.Performance/NBench.Tests.Performance.csproj]: Remind me: this change was largely made to recreate the same environment where we had seen these issues with Akka.NET, no?

File: build.sh (hunk 1 above)
Comment (heynickc) [inline build.sh]: This tells the .NET Core tooling to use Mono to build for those assemblies that target `net452`.  Otherwise it looks in what would be the windows directories for msbuild.exe (I think).  I'm not entirely sure if the build was ever happening on Mono/.NET4.5, we may have only been running the build on Linux for the .NET Core target (so we could ship and then come back to it for CI purposes).  That's been changed when I removed all the selective including/excluding projects in build.fsx

File: src/NBench/Sdk/Compiler/IDiscovery.cs (hunk 3 above)
Comment (heynickc) [inline src/NBench/Sdk/Compiler/IDiscovery.cs]: That was something I didn't really want to do, but I ended up feeling like I had to and maybe you have some more insight on it.  Before this, we would return a single `Assembly` from `LoadAssembly(string assemblyPath)`.  The result would get passed to `FindBenchmarks` above.  My assumption was that following this path, there needed to be the full set of assemblies required to compile the assembly and then find the benchmarks and execute them, which was what I was pinning as part of the root cause of our issue because it was here that the error was being thrown (implemented by `ReflectionDiscovery` class).

So instead of treating the `AssemblyLoadContext.Default` as a static container that held all of the dependencies, I loaded them all into this `Assembly[]` collection so that when work was done later to run the benchmarks, it had all the assemblies needed to do so, instead of the one single assembly.  This had the trickle effect for other methods that used `LoadAssembly` so I had to change their signature to accept an array of assemblies.

File: src/NBench/Sdk/Compiler/ReflectionDiscovery.Configurators.cs (hunk 4 above)
Comment (heynickc) [inline src/NBench/Sdk/Compiler/ReflectionDiscovery.Configurators.cs]: That was temporary and I forgot to change it back.  Was using it as a catch-all originally for the `LoadAssemblies` method on .NET Core.

File: tests/NBench.Tests.Performance/NBench.Tests.Performance.csproj (hunk 7 above)
Comment (heynickc) [inline tests/NBench.Tests.Performance/NBench.Tests.Performance.csproj]: Yes, and these benchmark assemblies aren't going to be ran as console apps so I figured it's fine to leave them as libraries instead of `netcoreapp1.1` targets although the runner should run against either targets.

File: tests/NBench.Tests.End2End/NBenchIntegrationTest.WithDependencies.cs (hunk 5 above)
Comment (heynickc) [inline tests/NBench.Tests.End2End/NBenchIntegrationTest.WithDependencies.cs]: Yep, will do.

File: tests/NBench.Tests.Performance.WithDependencies/Actor/ActorThroughputSpec.cs (hunk 6 above)
Comment (heynickc) [inline tests/NBench.Tests.Performance.WithDependencies/Actor/ActorThroughputSpec.cs]: Done
